File: backend/scoring/score_specialties.py
"""Python reference implementation — Deterministic specialty scoring.

Usage:
    python score_specialties.py

Reads from ../app/data/synonyms_tr.json and ../app/data/specialty_keywords_tr.json
"""

from __future__ import annotations
import json
import re
from dataclasses import dataclass, asdict
from pathlib import Path
from typing import Any, Dict, List, Set, Tuple

# Default paths (relative to this file)
_HERE = Path(__file__).resolve().parent
_DEFAULT_SYNONYMS = _HERE.parent / "app" / "data" / "synonyms_tr.json"
_DEFAULT_KEYWORDS = _HERE.parent / "app" / "data" / "specialty_keywords_tr.json"


def normalize_tr(text: str) -> str:
    """Turkish-aware normalization with proper İ/I case folding."""
    text = text.replace("\u0130", "i").replace("I", "\u0131")  # İ→i, I→ı
    text = text.lower().strip()
    text = re.sub(r"""[.,;:!?(){}\[\]"'`~]""", " ", text)
    text = re.sub(r"\s+", " ", text)
    return text


def load_json(path: str) -> Any:
    with open(path, "r", encoding="utf-8") as f:
        return json.load(f)


def unique(seq: List[str]) -> List[str]:
    seen: Set[str] = set()
    out: List[str] = []
    for x in seq:
        if x not in seen:
            out.append(x)
            seen.add(x)
    return out


@dataclass
class SpecialtyScore:
    id: str
    specialty_tr: str
    score: int
    phrase_score: int
    keyword_score: int
    negative_penalties: int
    matched_phrases_tr: List[str]
    matched_keywords_tr: List[str]
    matched_canonicals: List[str]
    debug: Dict[str, Any]


def build_synonym_index(syn: Dict[str, Any]) -> Tuple[List[Tuple[str, str]], Set[str]]:
    variants: List[Tuple[str, str]] = []
    for s in syn["synonyms"]:
        canonical = s["canonical"].lower()
        for v in s["variants_tr"]:
            variants.append((v.lower(), canonical))
    # longest first, then lexicographic for determinism
    variants.sort(key=lambda x: (-len(x[0]), x[0]))
    canonical_set = {s["canonical"].lower() for s in syn["synonyms"]}
    return variants, canonical_set


def score_specialties_deterministic(
    text_tr: str,
    synonyms_path: str = str(_DEFAULT_SYNONYMS),
    specialty_keywords_path: str = str(_DEFAULT_KEYWORDS),
) -> Dict[str, Any]:
    syn = load_json(synonyms_path)
    spec = load_json(specialty_keywords_path)

    variants, canonical_set = build_synonym_index(syn)
    normalized = normalize_tr(text_tr)

    matched_phrases: List[Tuple[str, str]] = []  # (phrase, canonical)
    canonical_locked: Set[str] = set()

    # 1) phrase detection
    for variant, canonical in variants:
        if variant in normalized:
            matched_phrases.append((variant, canonical))
            canonical_locked.add(canonical)

    # 2) canonical keywords from raw text (not locked already)
    matched_keywords: List[str] = []
    for canonical in sorted(canonical_set):  # sorted for determinism
        if canonical in normalized and canonical not in canonical_locked:
            matched_keywords.append(canonical)
            canonical_locked.add(canonical)

    phrase_canonicals = unique([c for _, c in matched_phrases])
    keyword_canonicals = unique(matched_keywords)

    keyword_points = int(spec["scoring"]["keyword_match_points"])
    phrase_points = int(spec["scoring"]["phrase_match_points"])
    neg_penalty = int(spec["scoring"]["negative_keyword_penalty"])

    scores: List[SpecialtyScore] = []

    for s in spec["specialties"]:
        sid = s["id"]
        sname = s["specialty_tr"]
        keywords = {k.lower() for k in s["keywords_tr"]}
        negatives = [n.lower() for n in s.get("negative_keywords_tr", [])]

        score = 0
        phrase_score = 0
        keyword_score = 0
        negative_penalties = 0
        scored_canonicals: Set[str] = set()
        matched_phr_out: List[str] = []
        matched_kw_out: List[str] = []
        hits: List[Dict[str, Any]] = []

        # phrase scoring (once per canonical)
        for canonical in phrase_canonicals:
            rep_phrase = next((p for p, c in matched_phrases if c == canonical), None)
            phrase_matches = (canonical in keywords) or (rep_phrase in keywords if rep_phrase else False)

            if phrase_matches and canonical not in scored_canonicals:
                score += phrase_points
                phrase_score += phrase_points
                scored_canonicals.add(canonical)
                if rep_phrase:
                    matched_phr_out.append(rep_phrase)
                hits.append({"kind": "phrase", "value": rep_phrase or canonical, "points": phrase_points})

        # keyword scoring (canonicals not already scored)
        for canonical in keyword_canonicals:
            if canonical in keywords and canonical not in scored_canonicals:
                score += keyword_points
                keyword_score += keyword_points
                scored_canonicals.add(canonical)
                matched_kw_out.append(canonical)
                hits.append({"kind": "keyword", "value": canonical, "points": keyword_points})

        # negative penalties
        for neg in negatives:
            if neg and neg in normalized:
                score += neg_penalty
                negative_penalties += neg_penalty
                hits.append({"kind": "negative", "value": neg, "points": neg_penalty})

        scores.append(SpecialtyScore(
            id=sid,
            specialty_tr=sname,
            score=score,
            phrase_score=phrase_score,
            keyword_score=keyword_score,
            negative_penalties=negative_penalties,
            matched_phrases_tr=unique(matched_phr_out),
            matched_keywords_tr=unique(matched_kw_out),
            matched_canonicals=sorted(scored_canonicals),
            debug={"normalized_text": normalized, "hits": hits},
        ))

    # deterministic ordering
    scores.sort(key=lambda x: (-x.score, -x.keyword_score, x.id))

    tie = False
    if len(scores) > 1 and scores[0].score == scores[1].score and scores[0].keyword_score == scores[1].keyword_score:
        tie = True

    return {
        "normalized_text": normalized,
        "top": asdict(scores[0]) if scores else None,
        "scores": [asdict(s) for s in scores],
        "tie": tie,
    }


if __name__ == "__main__":
    test_cases = [
        "Sabah kalktığımdan beri başım dönüyor, midem bulanıyor.",
        "3 gündür boğazım ağrıyor, ateşim var, yutkunurken yanıyor.",
        "Göğsümde baskı var, nefesim daralıyor ve terliyorum.",
        "İdrar yaparken yanıyor, çok sık tuvalete çıkıyorum.",
        "Vücudumda kaşıntılı döküntü çıktı.",
    ]

    for text in test_cases:
        res = score_specialties_deterministic(text)
        top = res["top"]
        print(f"\nInput: {text}")
        print(f"  Top: {top['specialty_tr']} (score={top['score']}, tie={res['tie']})")
        for s in res["scores"][:3]:
            if s["score"] > 0:
                print(f"    {s['specialty_tr']}: {s['score']}")

File: src/scoring/scoreSpecialties.ts
/* eslint-disable @typescript-eslint/no-explicit-any */
/**
 * TypeScript Reference Implementation — Deterministic Specialty Scoring
 *
 * Usage (Node.js):
 *   import { scoreSpecialtiesDeterministic } from './scoreSpecialties';
 *   const res = scoreSpecialtiesDeterministic(
 *     "Başım dönüyor, midem bulanıyor",
 *     "./config/synonyms_tr.json",
 *     "./config/specialty_keywords_tr.json"
 *   );
 *   console.log(res.top);
 */
import fs from "fs";

type SynonymsFile = {
  version: string;
  language: string;
  rules: {
    double_count_policy: "NO_DOUBLE_COUNT_SAME_CANONICAL";
    canonical_priority: ("phrase" | "keyword")[];
  };
  synonyms: Array<{
    canonical: string;
    type: "symptom" | "red_flag";
    variants_tr: string[];
  }>;
};

type SpecialtyKeywordsFile = {
  version: string;
  language: string;
  specialties: Array<{
    id: string;
    specialty_tr: string;
    keywords_tr: string[];
    negative_keywords_tr: string[];
  }>;
  scoring: {
    keyword_match_points: number; // 3
    phrase_match_points: number;  // 5
    negative_keyword_penalty: number; // -4
    tie_breakers: string[];
  };
};

export type SpecialtyScore = {
  id: string;
  specialty_tr: string;
  score: number;
  phrase_score: number;
  keyword_score: number;
  negative_penalties: number;
  matched_phrases_tr: string[];
  matched_keywords_tr: string[];
  matched_canonicals: string[]; // scored canonicals
  debug: {
    normalized_text: string;
    hits: Array<{ kind: "phrase" | "keyword" | "negative"; value: string; points: number }>;
  };
};

function normalizeTR(input: string): string {
  // deterministic normalization with Turkish İ/I case folding
  return input
    .replace(/\u0130/g, "i")   // İ → i (Turkish capital I with dot)
    .replace(/I/g, "\u0131")   // I → ı (Turkish capital I without dot)
    .toLowerCase()
    .trim()
    .replace(/[.,;:!?(){}\[\]"'`~]/g, " ")
    .replace(/\s+/g, " ");
}

function unique<T>(arr: T[]): T[] {
  return Array.from(new Set(arr));
}

function loadJson<T>(path: string): T {
  return JSON.parse(fs.readFileSync(path, "utf-8")) as T;
}

export function buildSynonymIndex(syn: SynonymsFile) {
  // variants sorted by length desc for deterministic longest-match-first
  const variants: Array<{ variant: string; canonical: string; type: string }> = [];
  for (const s of syn.synonyms) {
    for (const v of s.variants_tr) {
      variants.push({ variant: v.toLowerCase(), canonical: s.canonical.toLowerCase(), type: s.type });
    }
  }
  variants.sort((a, b) => b.variant.length - a.variant.length || a.variant.localeCompare(b.variant));

  // canonical set for quick keyword checks
  const canonicalSet = new Set(syn.synonyms.map(s => s.canonical.toLowerCase()));
  return { variants, canonicalSet };
}

export function scoreSpecialtiesDeterministic(
  text_tr: string,
  synonymsPath: string,
  specialtyKeywordsPath: string
): { normalized_text: string; scores: SpecialtyScore[]; top: SpecialtyScore; tie: boolean } {
  const syn = loadJson<SynonymsFile>(synonymsPath);
  const spec = loadJson<SpecialtyKeywordsFile>(specialtyKeywordsPath);
  const { variants, canonicalSet } = buildSynonymIndex(syn);

  const normalized = normalizeTR(text_tr);

  // 1) Phrase detection -> matched canonicals (NOT scored yet)
  const matchedPhrases: Array<{ phrase: string; canonical: string }> = [];
  const canonicalLocked = new Set<string>(); // NO_DOUBLE_COUNT_SAME_CANONICAL

  for (const item of variants) {
    if (normalized.includes(item.variant)) {
      // if canonical already locked, still allow storing phrase for UX,
      // but do not create new canonical
      matchedPhrases.push({ phrase: item.variant, canonical: item.canonical });
      canonicalLocked.add(item.canonical);
    }
  }

  // 2) Canonical keywords appearing directly in text (if not already locked by phrase)
  const matchedKeywords: string[] = [];
  for (const canonical of Array.from(canonicalSet).sort()) {
    if (normalized.includes(canonical) && !canonicalLocked.has(canonical)) {
      matchedKeywords.push(canonical);
      canonicalLocked.add(canonical);
    }
  }

  // For scoring, we want: phrase contributes once per canonical (priority phrase>keyword)
  // We'll compute scoredCanonicals from phrases first, then keywords.
  const phraseCanonicalsOrdered = unique(matchedPhrases.map(p => p.canonical));
  const keywordCanonicalsOrdered = unique(matchedKeywords);

  // Precompute per-specialty keyword sets (lowercase)
  const specialtyData = spec.specialties.map(s => ({
    id: s.id,
    specialty_tr: s.specialty_tr,
    keywords: new Set(s.keywords_tr.map(k => k.toLowerCase())),
    negatives: s.negative_keywords_tr.map(n => n.toLowerCase())
  }));

  const scores: SpecialtyScore[] = [];

  for (const s of specialtyData) {
    let score = 0;
    let phraseScore = 0;
    let keywordScore = 0;
    let negativePenalties = 0;

    const matched_phrases_tr: string[] = [];
    const matched_keywords_tr: string[] = [];
    const scoredCanonicals = new Set<string>();
    const hits: SpecialtyScore["debug"]["hits"] = [];

    // Phrase scoring: if canonical OR phrase literal exists in specialty keywords
    for (const canonical of phraseCanonicalsOrdered) {
      // Find a representative phrase for UX
      const anyPhrase = matchedPhrases.find(p => p.canonical === canonical)?.phrase;

      const phraseMatchesSpecialty =
        s.keywords.has(canonical) || (anyPhrase ? s.keywords.has(anyPhrase) : false);

      if (phraseMatchesSpecialty && !scoredCanonicals.has(canonical)) {
        score += spec.scoring.phrase_match_points;
        phraseScore += spec.scoring.phrase_match_points;
        scoredCanonicals.add(canonical);
        if (anyPhrase) matched_phrases_tr.push(anyPhrase);
        hits.push({ kind: "phrase", value: anyPhrase ?? canonical, points: spec.scoring.phrase_match_points });
      }
    }

    // Keyword scoring (only canonicals not scored by phrase)
    for (const canonical of keywordCanonicalsOrdered) {
      if (s.keywords.has(canonical) && !scoredCanonicals.has(canonical)) {
        score += spec.scoring.keyword_match_points;
        keywordScore += spec.scoring.keyword_match_points;
        scoredCanonicals.add(canonical);
        matched_keywords_tr.push(canonical);
        hits.push({ kind: "keyword", value: canonical, points: spec.scoring.keyword_match_points });
      }
    }

    // Negative penalties
    for (const neg of s.negatives) {
      if (neg && normalized.includes(neg)) {
        score += spec.scoring.negative_keyword_penalty;
        negativePenalties += spec.scoring.negative_keyword_penalty;
        hits.push({ kind: "negative", value: neg, points: spec.scoring.negative_keyword_penalty });
      }
    }

    scores.push({
      id: s.id,
      specialty_tr: s.specialty_tr,
      score,
      phrase_score: phraseScore,
      keyword_score: keywordScore,
      negative_penalties: negativePenalties,
      matched_phrases_tr: unique(matched_phrases_tr),
      matched_keywords_tr: unique(matched_keywords_tr),
      matched_canonicals: Array.from(scoredCanonicals).sort(),
      debug: { normalized_text: normalized, hits }
    });
  }

  // Deterministic sorting + tie flag
  scores.sort((a, b) => {
    if (b.score !== a.score) return b.score - a.score;
    if (b.keyword_score !== a.keyword_score) return b.keyword_score - a.keyword_score; // tie-break 1
    return a.id.localeCompare(b.id); // deterministic final tie-break
  });

  const top = scores[0];
  const tie = scores.length > 1 && scores[0].score === scores[1].score && scores[0].keyword_score === scores[1].keyword_score;

  return { normalized_text: normalized, scores, top, tie };
}
